pennylane/circuit_graph.py

```diff
@@ -302,14 +257,25 @@ def ancestors(self, ops):
         Returns:
             list[Operator]: ancestors of the given operators
         """
-        # rx.ancestors() returns node indices instead of node-values
-        all_indices = set().union(*(rx.ancestors(self._graph, self._indices[id(o)]) for o in ops))
-        double_op_indices = set(self._indices[id(o)] for o in ops)
-        ancestor_indices = all_indices - double_op_indices
-
-        return list(self._graph.get_node_data(n) for n in ancestor_indices)
-
-    def descendants(self, ops):
+        if isinstance(ops, (Operator, MeasurementProcess)):
+            raise ValueError(
+                "CircuitGraph.ancestors accepts an iterable of"
+                " operators and measurements, not operators and measurements themselves."
+            )
+        if any(len(self._inds_for_objs[WrappedObj(op)]) > 1 for op in ops):
+            raise ValueError(
+                "Cannot calculate ancestors for an operator that occurs multiple times."
+            )
+        ancestors = set()
+        for op in ops:
+            ind = self._inds_for_objs[WrappedObj(op)][0]
+            op_ancestors = rx.ancestors(self._graph, ind)
+            ancestors.update(set(op_ancestors))
+        if sort:
+            ancestors = sorted(ancestors)
+        return [self._queue[ind] for ind in ancestors]
+
+    def descendants(self, ops, sort=False):
```

Comment:
same thing about allowing us to reduce the duplication with `descendants_in_order`.